File: python/demo18a01_barrier.py
'''
BARRIERS AND LATCHES
Version A: Barriers
'''

import time
import threading



sync_point = threading.Barrier(parties=3)



def process_request(user_name: str, time_wait: int):
    time.sleep(time_wait)

    print(f'Get request from {user_name}')
    sync_point.wait()

    print(f'Process request for {user_name}')
    sync_point.wait()

    print(f'Done {user_name}')



lstarg = [
    ('lorem', 1),
    ('ipsum', 2),
    ('dolor', 3)
]

_ = [ threading.Thread(target=process_request, args=arg).start() for arg in lstarg ]

File: csharp/demo/demo18a01-barrier.cs
/*
 * BARRIERS AND LATCHES
 * Version A: Barriers
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo18A01 : IRunnable
{
    public void run()
    {
        var syncPoint = new Barrier(participantCount: 3);


        var lstArg = new List<ThreadArg>
        {
            new ThreadArg{ userName = "lorem", waitTime = 1 },
            new ThreadArg{ userName = "ipsum", waitTime = 2 },
            new ThreadArg{ userName = "dolor", waitTime = 3 }
        };


        lstArg.ForEach(arg => new Thread(() =>
        {

            Thread.Sleep(1000 * arg.waitTime);

            Console.WriteLine("Get request from " + arg.userName);
            syncPoint.SignalAndWait();

            Console.WriteLine("Process request for " + arg.userName);
            syncPoint.SignalAndWait();

            Console.WriteLine("Done " + arg.userName);

        }).Start());
    }



    class ThreadArg
    {
        public string userName;
        public int waitTime;
    }
}
